### PR TITLE
Allow guests to operate in encrypted rooms

### PR DESCRIPTION
**Pull request by awesome-michael**
Guests are not able to operate in encrypted rooms and keep getting lots of errors. These changes enable the guests to write in encrypted rooms. I think one problem is that the guest was not allowed to get the member list of the room and therefore had no list which encryption keys he has to request to encrypt his message properly.
Honestly I'm not sure why I needed the change in the events.py file. I tested this configuration and guests can write and read in encrypted rooms now.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Signed-off-by: Michael Albert <michael.albert@awesome-technologies.de>

### CODE REVIEW

**Comment by clokep** [comment]
@awesome-michael Looks like this needs a newsfile added (and maybe a merge from develop since it has been a bit)? Were you waiting for feedback on the approach?

**Comment by awesome-michael** [comment]
Is a newsfile the same as the changelog mentioned in the checklist?
Do you know if these endpoints were blocked for guests on purpose?

File: synapse/rest/client/v1/events.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def __init__(self, hs):
         self._event_serializer = hs.get_event_client_serializer()
 
     async def on_GET(self, request, event_id):
-        requester = await self.auth.get_user_by_req(request)
+        requester = await self.auth.get_user_by_req(request, allow_guest=True)
```
Comment (richvdh) [inline synapse/rest/client/v1/events.py:84]: this endpoint is deprecated, and nothing should be using it. See https://matrix.org/docs/spec/client_server/r0.6.1#deprecated-get-matrix-client-r0-events-eventid.

**Comment by richvdh** [comment]
> Is a newsfile the same as the changelog mentioned in the checklist?

yes.

>  Do you know if these endpoints were blocked for guests on purpose?

The list of endpoints which guests can access is specced at https://matrix.org/docs/spec/client_server/r0.6.1#id112. If we want to open up more endpoints, then that is a spec change, which would require a spec-change proposal.

I'm a bit surprised if `/rooms/<room_id>/members` is the only thing that needs changing to allow guest users to operate in encrypted rooms, but if that's true then certainly it seems to make sense to change the specification and synapse in this way.

**Comment by richvdh** [comment]
@awesome-manuel just checking in on this. are you still interested in working on it?

**Comment by awesome-michael** [comment]
From your last comment I assume there is a spec change needed before this PR can be used. Unfortunately I'm not sure I can write a propper spec-change proposal for that as I'm not enough into the details of the existing guest concept.

**Comment by richvdh** [comment]
Ok, thanks. My impression is that having guests operate in e2e rooms is a fundamentally flawed concept, since there is no way for them to get the keys to the events in the room (which is probably as it should be: there's no point in encrypting events if anyone can turn up as a guest and get the keys to those events).

So for now I'm going to close this, but if you'd like to discuss this in more detail, we can!

**Comment by awesome-michael** [comment]
I kind of disagree. If you dont want guests to join the room thats what the guest join rules are for. Now you force guests (and users that want to communicate with them) to use unencrypted rooms.
How would the keys to event be retrieved (as the events endpoint is deprecated as you pointed out above)?

**Comment by ptman** [comment]
The chat is probably still encrypted (HTTPS), just not e2e encrypted. The hard problem with encryption is usually key management. How do you manage keys with a guest account?

**Comment by awesome-manuel** [comment]
Maybe we should explain our use case in more detail:
- We need user accounts that can be invited (e.g. via a link) into a 1-to-1 chat or into a room
- Those users are not allowed to register without an invitation
- Those users are not allowed to create new rooms on their own

Especially the last point is tricky to enforce with normal user accounts, so we had the idea to use guest accounts for that. The guest accounts are only for temporary conversations and all content can or must be deleted after that conversation.

**Comment by richvdh** [comment]
Ok, this is fair, and I mis-spoke when I said that this was a fundamentally flawed concept; I'd forgotten that guest users could actually join rooms in the same way as normal users.

Now that I look again at the list of what guest users can and can't do, I can't actually see much reason that e2e shouldn't work for them. 

So a couple of things here: firstly, have you actually tested this, and confirmed that it is both necessary and sufficient to get e2e working correctly for guest accounts? If so, I'd very much encourage you to create an MSC for this: it shouldn't need much more than creating a document that explains the proposed change and why it is a useful extension to the protocol.

However, I'd also say this: it's unclear to me what role guest accounts actually have in the matrix protocol and where the lines of what you can and can't do end up. That's a larger topic which might bear discussion in `#matrix-spec:matrix.org` or similar - I don't think this PR is the best place.

> Especially the last point is tricky to enforce with normal user accounts,

Have you looked into third_party_event_rules? https://github.com/matrix-org/synapse/blob/master/docs/sample_config.yaml#L2163. It provides a hook where you can say whether a given user is, or is not, allowed to create a room.

**Comment by awesome-michael** [comment]
Now that the MSC is merged, would you open this PR again or should I file a new one?

**Comment by clokep** [comment]
@awesome-michael I think you'll need to merge in `develop` to get CI to pass. 👍

**Comment by anoadragon453** [comment]
@awesome-michael Your changelog must end in either `.doc`, `.bugfix`, `.misc`, `.feature`, `.removal` or `.docker`. Please see the [contributing guide](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog) on this for more info.

For this change, I would probably mark it as a `.misc` (internal change) or even a `.feature`. I don't think I would consider it a bug fix.

**Comment by awesome-michael** [comment]
Thanks for that hint.. the error message didn't help me getting there..

**Pull request merged**